url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/model3d/SegmentNet3DComp.java

```diff
@@ -0,0 +1,1307 @@
+	 * the "left" neighbour of the EntryElement.
+	 * 
+	 * @return end - SegmentElt3D, the end element.
+	 */
+	public SegmentElt3D getEndElement() {// Dag
+		if (this.isClosed())
+			return (this.getEntryElement().getNeighbour(1));
+		SegmentElt3D current = this.getEntryElement();
+		while (current.hasNeighbour(0)) {
+			current = current.getNeighbour(0);
+		}
+		return current;
+	}
+
+	/**
+	 * Adds the given element to the component.<br>
+	 * The given element reference is not valid anymore after insertion.<br>
+	 * If you need to hold a reference on the element, update your variable with
+	 * the return value element !
+	 * 
+	 * @param elt
+	 *            Segment3D to be added
+	 * @return SegmentElt3D - the inserted instance.
+	 * @throws UpdateException
+	 *             - or subclass of it, signals an Update problem.
+	 * @throws IllegalStateException
+	 *             - if the intersectsInt(Line3D line, ScalarOperator sop)
+	 *             method of the class Line3D (which computes the intersection
+	 *             of two lines) called by this method returns a value that is
+	 *             not -2, -1, 0 or 1.
+	 * @throws IllegalArgumentException
+	 *             - if an attempt is made to construct a MBB3D whose maximum
+	 *             point is not greater than its minimum point.
+	 * @throws ArithmeticException
+	 *             - if norm equals zero in epsilon range. This exception
+	 *             originates in the method normalize(ScalarOperator) of the
+	 *             class Vector3D.
+	 */
+	public SegmentElt3D addElt(Segment3D elt) throws UpdateException { // Dag
+		if (this.containsElt(elt))
+			throw new ContainmentException("Element already contained !");
+
+		return addElt(elt.getPoint(0), elt.getPoint(1));
+	}
+
+	/**
+	 * Creates a new element and adds it to the component. One of the points p1
+	 * or p2 must be geometry equivalent to an end point of the net.
+	 * 
+	 * @param p1
+	 *            Point3D, start point of the SegmentElt3D that should be
+	 *            created
+	 * @param p2
+	 *            Point3D, end point of the SegmentElt3D that should be created
+	 * @return SegmentElt3D - inserted element.
+	 * @throws UpdateException
+	 *             - or subclass of it, signals an Update problem.
+	 * @throws IllegalArgumentException
+	 *             - if an attempt is made to construct a MBB3D whose maximum
+	 *             point is not greater than its minimum point.
+	 * @throws IllegalArgumentException
+	 *             - if index of a triangle point is not 0, 1 or 2. The
+	 *             exception originates in the method getPoint(int) of the class
+	 *             Triangle3D.
+	 * @throws ArithmeticException
+	 *             - if norm equals zero in epsilon range. This exception
+	 *             originates in the method normalize(ScalarOperator) of the
+	 *             class Vector3D.
+	 */
+	public SegmentElt3D addElt(Point3D p1, Point3D p2) throws UpdateException { // Dag
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 418-428 into a separate method.